### PR TITLE
feat(tabs): update to scrollable tabs

### LINKED ISSUE #1724: Support for scrolling tabs with overflow nav buttons
> Describe in detail the issue you're having.
Currently when tabs overflow, it changes to a drop down, something that can be replicated by shrinking the window size here: https://angular.carbondesignsystem.com/?path=/story/components-tabs--basic

We have a requirement for our tabs to support the carbon components style scrolling with overflow nav buttons.
Something that has been working for the react framework, discussed here https://github.com/carbon-design-system/carbon/pull/6410

> Is this a feature request (new component, new icon), a bug, or a general issue?
It's a feature request to replace the drop down or add support for scrollable.

> Is this issue related to a specific component?
Tabs

> What browser are you working in?
Chrome

> What version of the Carbon Design System are you using?
carbon-components: v10.27.0 
carbon-components-angular: v4.39

> What offering/product do you work on? Any pressing ship or release dates we should be aware of?
We are working on a MVP deadline and want to know if you have plans to remedy this in the new feature and could provide ETA?

### PR DESCRIPTION
**Pull request by zvonimirfras**
Closes #1724 

Brings the tabs up to date with react implementation. It borrows some approaches and interfaces from react implementation to match the behavior as closely as possible.

Tabs could use a bit of refactoring. Because of differences between tab headers and tab header group, it might not be reasonable to have one extend the other. It might make sense to have a tab-headers-base that contains common functionality between the two, that both extend. Not in this PR though. If we agree on this approach, I'll open an issue with this info.

### CODE REVIEW

**Review by cal-smith — APPROVED** [review]
Looks good to me!

> Tabs could use a bit of refactoring. Because of differences between tab headers and tab header group, it might not be reasonable to have one extend the other. It might make sense to have a tab-headers-base that contains common functionality between the two, that both extend. Not in this PR though. If we agree on this approach, I'll open an issue with this info.

I would be inclined to take a bit of a different approach - if we're going to do a major refactor, we should decompose the tabs into a bunch of components and directives and use that to build up whatever variants we need. The scrolling behaviour could be a directive, the `nav` container a component, the header items a component, etc.

File: src/tabs/tab-header-group.component.ts [hunk 1]
```diff
@@ -104,20 +128,47 @@ export class TabHeaderGroup implements AfterContentInit, OnDestroy, OnChanges {
 	 * ContentChildren of all the tabHeaders.
 	 */
 	@ContentChildren(TabHeader) tabHeaderQuery: QueryList<TabHeader>;
+	// @ts-ignore
```
[the remark this answers is not in the document]
Comment (zvonimirfras) [reply src/tabs/tab-header-group.component.ts:131]:
```
Argument of type '{ static: boolean; }' is not assignable to parameter of type '{ read?: any; }'.
Object literal may only specify known properties, and 'static' does not exist in type '{ read?: any; }'
```

**Comment by zvonimirfras** [comment]
> Looks good to me!
> 
> > Tabs could use a bit of refactoring. Because of differences between tab headers and tab header group, it might not be reasonable to have one extend the other. It might make sense to have a tab-headers-base that contains common functionality between the two, that both extend. Not in this PR though. If we agree on this approach, I'll open an issue with this info.
> 
> I would be inclined to take a bit of a different approach - if we're going to do a major refactor, we should decompose the tabs into a bunch of components and directives and use that to build up whatever variants we need. The scrolling behaviour could be a directive, the `nav` container a component, the header items a component, etc.

I did say "a bit" 😄 
If we're doing a major refactor, it might as well be all of it.

File: src/tabs/tab-headers.component.ts [hunk 2]
```diff
@@ -281,86 +329,75 @@ export class TabHeaders implements AfterContentInit, OnChanges {
...
+			direction === 1 &&
+			scrollLeft + clientWidth >= scrollWidth - this.OVERFLOW_BUTTON_OFFSET;
+		if (leftEdgeReached || rightEdgeReached) {
+			if (leftEdgeReached) {
+				this.rightOverflowNavButton.nativeElement.focus();
+			}
+			if (rightEdgeReached) {
+				this.leftOverflowNavButton.nativeElement.focus();
+			}
 		}
```
Comment (Donisius) [inline src/tabs/tab-headers.component.ts:375]: Is the parent `leftEdgeReached || rightEdgeReached`  necessary here?

File: src/tabs/tab-header-group.component.ts [hunk 3]
```diff
@@ -259,67 +304,62 @@ export class TabHeaderGroup implements AfterContentInit, OnDestroy, OnChanges {
...
-		if (!this.tabListVisible) { return; }
+		tabList.scrollLeft += direction * multiplier;
 
-		const target = (event.target as HTMLElement).closest("a");
+		const leftEdgeReached =
+			direction === -1 && scrollLeft < this.OVERFLOW_BUTTON_OFFSET;
+		const rightEdgeReached =
+			direction === 1 &&
+			scrollLeft + clientWidth >= scrollWidth - this.OVERFLOW_BUTTON_OFFSET;
+		if (leftEdgeReached || rightEdgeReached) {
```
Comment (Donisius) [inline src/tabs/tab-header-group.component.ts:326]: Is the parent `leftEdgeReached || rightEdgeReached` necessary here?

**Review by Donisius — CHANGES_REQUESTED** [review]

File: src/tabs/tab-headers.component.ts (hunk 2 above)
Comment (zvonimirfras) [reply src/tabs/tab-headers.component.ts:375]: You know what? Probably not. This logic comes from the reference implementation.

**Review by Donisius — APPROVED** [review]

File: src/tabs/tab-header-group.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (cal-smith) [reply src/tabs/tab-header-group.component.ts:131]: Huh ... odd

File: src/tabs/tab-header-group.component.ts [hunk 4]
```diff
@@ -191,18 +242,12 @@ export class TabHeaderGroup implements AfterContentInit, OnDestroy, OnChanges {
...
 
-	@HostListener("focusout", ["$event"])
-	focusOut(event) {
-		if (this.tabListVisible && !this.elementRef.nativeElement.contains(event.relatedTarget)) {
-			this.tabListVisible = false;
-		}
+	ngOnInit() {
+		this.windowResizeSubscription = fromEvent(window, "resize", { passive: true })
+			.pipe(debounceTime(200))
+			.subscribe(() => this.handleScroll());
```
Comment (cal-smith) [inline src/tabs/tab-header-group.component.ts:250]: Would the `DocumentService` or `EventService` in `utils/UtilsModule` help here? Both of them automatically handle and clean up subscriptions, which would make this a bit cleaner

Comment (cal-smith) [reply src/tabs/tab-header-group.component.ts:250]: No debounce, but honestly all that usually does is slow down the UI - most changes are fast enough without a debounce that it feels way better.

File: src/tabs/tab-headers.component.ts [hunk 5]
```diff
@@ -232,18 +281,12 @@ export class TabHeaders implements AfterContentInit, OnChanges {
...
 
-	@HostListener("focusout", ["$event"])
-	focusOut(event) {
-		if (this.tabListVisible && !this.elementRef.nativeElement.contains(event.relatedTarget)) {
-			this.tabListVisible = false;
-		}
+	ngOnInit() {
+		this.windowResizeSubscription = fromEvent(window, "resize", { passive: true })
+			.pipe(debounceTime(200))
+			.subscribe(() => this.handleScroll());
```
Comment (cal-smith) [inline src/tabs/tab-headers.component.ts:289]: Same comment here with regards to the event handling

**Review by cal-smith — CHANGES_REQUESTED** [review]
Overall looks good, just one small thing I missed the first time around

File: src/tabs/tab-header-group.component.ts (hunk 4 above)
Comment (zvonimirfras) [reply src/tabs/tab-header-group.component.ts:250]: The debounce makes it function equivalently to reference implementation.

Comment (cal-smith) [reply src/tabs/tab-header-group.component.ts:250]: Do we need it though? Playing with the react implementation, it definitely makes it a bit weird (the overflow button hangs around for a while after it should disappear), and 200ms is pretty short - how much does that really improve performance?

Either way, if we do need it we can adjust our services as needed, and my original question is still relevant: can we use `DocumentService` or `EventService` for the resize events? Auto-cleanup is a pretty compelling bonus I think.

Comment (zvonimirfras) [reply src/tabs/tab-header-group.component.ts:250]: From a usability perspective, `EventService` takes gold compared to `fromEvent`

**Review by Donisius — APPROVED** [review]

**Review by cal-smith — APPROVED** [review]
Nice! Everything seems good to me 🎉

**Pull request merged**